### ISSUE TITLE
ValueCollection not saving values

### ISSUE THREAD

**Issue by joaolisboa** [rinvex/laravel-attributes#47]
I'm having trouble working with the ValueCollection. The documentation doesn't help because it doesn't specify in any way how the ValueCollection is supposed to be used for multi-value attributes other than how to store the values with push() and add(). I thought it'd work by creating a custom attribute but instead of Value it'd instance ValueCollection but Laravel throws an undefined Relation error. Can you help with undestanding the ValueCollection?

**Comment by joaolisboa** [comment]
Well, nevermind my stupidity since there's a collection property for when creating attributes. I still have an issue though.

Testing in artisan tinker, and when I push a value and save the model, it doesn't update the collection. Example output from tinker:


    >>> $product = Modules\Catalog\Models\Product::find(2);
    => Modules\Catalog\Models\Product {#3154
         id: 2,
         sku: "test",
         name: "test",
         family_id: 1,
         product_brand_id: 1,
         product_state_id: 1,
         created_at: null,
         updated_at: null,
         deleted_at: null,
         product_segment_id: 1,
         product_manager: 1,
         photo: null,
         attribute-1: Rinvex\Attributes\Models\Type\Integer {#3184
           id: 1,
           content: 23,
           attribute_id: 21,
           entity_id: 2,
           entity_type: "Modules\Catalog\Models\Product",
           created_at: "2018-06-06 23:46:22",
           updated_at: "2018-06-06 23:46:22",
         },
         attribute-1-fixed: null,
         attribute-1-1: null,
         description: Rinvex\Attributes\Models\Type\Varchar {#3194
           id: 1,
           content: "descrição produto #2",
           attribute_id: 24,
           entity_id: 2,
           entity_type: "Modules\Catalog\Models\Product",
           created_at: null,
           updated_at: null,
         },
         file: null,
         file-1: null,
         option: App\Models\Option {#3216
           id: 1,
           content: 2,
           attribute_id: 27,
           entity_id: 2,
           entity_type: "Modules\Catalog\Models\Product",
           created_at: "2018-06-06 23:46:22",
           updated_at: "2018-06-06 23:46:22",
         },
         options: Rinvex\Attributes\Support\ValueCollection {#3228
           all: [
             App\Models\Options {#3229
               id: 1,
               content: 1,
               attribute_id: 29,
               entity_id: 2,
               entity_type: "Modules\Catalog\Models\Product",
               created_at: "2018-06-07 00:08:39",
               updated_at: "2018-06-07 00:08:39",
             },
           ],
         },
       }


    >>> $product->options->push("2")
    => Illuminate\Support\Collection {#3077
         all: [
           "1",
           "2",
         ],
       }


    >>> $product->options
    => Illuminate\Support\Collection {#3141
         all: [
           "1",
         ],
       }


    >>> $product->save()
    => true

    >>> $product->options
    => Illuminate\Support\Collection {#3225
         all: [
           "1",
         ],
       }

**Comment by uphlewis** [comment]
This issue appears to be in the same vein as [#70](https://github.com/rinvex/laravel-attributes/issues/70).

If you look closely at the returned collection when trying to access the ->options property, its a plain `Illuminate\Support\Collection` object, rather than an instance of the extended class `Rinvex\Attributes\Support\ValueCollection`.

I believe this is because when you aren't physically assigning a value to an attribute such as in your case, PHP is calling `__get()` (which calls `->getAttribute()` on the eloquent model) and not `__set()` (which calls `->setAttribute()` on the eloquent model). Because it's calling __get() rathern than __set() 
 the package model trait `Rinvex\Attributes\Trait\Attributable` is returning the `content` of the ValueCollection, which is just a standard Collection, rather than the ValueCollection itself.

Take a look at `Rinvex\Attributes\Trait\Attributable::getEntityAttributeValue()` and see for yourself: https://github.com/rinvex/laravel-attributes/blob/v1.0.0/src/Traits/Attributable.php#L310

This means you don't have access to `ValueCollection::add()`, so its natural to reach for `Collection::push()` because it doesnt throw the fatal error described in #70 however because Collections are immutable, `push()` returns a new instance of the Collection rather than modifying the existing instance, so to get it to work you'd need to assign the returned collection back to the model.

**tldr:**

So to get it to work you currently have 3 options AFAIK:
1. `$product->options = collect(['1', '2']); // assign an entire collection (or array etc) to the options property`
2. `$product->options = $product->options->push('2'); // assign the newly 'pushed' collection to the options property`
3. `$product->getRelation('options')->add('2'); //obtain the ValueCollection instance and use the custom add() method`

Followed by `$product->save();`

**Comment by Omranic** [comment]
This has been completely refactored in a different branch here:
https://github.com/rinvex/laravel-attributes/tree/refactor-to-native-laravel-relationships

In the new refactor, it actually uses normal relationships, and should be easy and straightforward like default Laravel relationships. Although, that refactor is incomplete. Any help with that branch would be much appreciated! 🙂
